File: study-plan-data-structure-1/array/max_subarray.py
from typing import List


class Solution:
    # Runtime: 716 ms, faster than 5.88% of Python3 online submissions for Maximum Subarray.
    # Memory Usage: 28.7 MB, less than 5.41% of Python3 online submissions for Maximum Subarray.
    def maxSubArray(self, nums: List[int]) -> int:
        max_sum = nums[0]
        cur_sum = 0

        for num in nums:
            if cur_sum < 0:
                cur_sum = 0
            cur_sum += num
            max_sum = max(max_sum, cur_sum)

        return max_sum

File: study-plan-data-structure-1/src/arrays/MaxSubArray.java
package arrays;

/**
 * 53. Maximum Subarray (easy)
 * <p>
 * Given an integer array nums, find the contiguous subarray (containing at least one number) which has the largest sum and return its sum.
 * A subarray is a contiguous part of an array.
 */
public class MaxSubArray {

    /**
     * Runtime: 1 ms, faster than 66.66% of Java online submissions for Maximum Subarray.
     * Memory Usage: 49.3 MB, less than 7.06% of Java online submissions for Maximum Subarray.
     */
    public int maxSubArray(int[] nums) {
        int maxSum = nums[0];
        int currentSum = 0;

        for (int num : nums) {
            if (currentSum < 0)
                currentSum = 0;

            currentSum += num;
            maxSum = Math.max(maxSum, currentSum);
        }

        return maxSum;
    }
}
